### ISSUE TITLE
Crash on train-segment

### ISSUE THREAD

**Issue by johnml1135** [sillsdev/machine#142]
Here are the logs:
```
2023-11-28 02:57:35	
machine-engine: 2023-11-28T07:57:31.956844993Z info: Microsoft.Hosting.Lifetime[0]
      Content root path: /app/engine_server/
2023-11-28 02:57:32	
machine-engine: 2023-11-28T07:57:31.956781112Z info: Microsoft.Hosting.Lifetime[0]
      Hosting environment: Production
2023-11-28 02:57:32	
machine-engine: 2023-11-28T07:57:31.956756852Z info: Microsoft.Hosting.Lifetime[0]
      Application started. Press Ctrl+C to shut down.
2023-11-28 02:57:32	
machine-engine: 2023-11-28T07:57:31.956059779Z info: Microsoft.Hosting.Lifetime[14]
      Now listening on: http://[::]:80
2023-11-28 02:57:32	
machine-engine: 2023-11-28T07:57:31.939813561Z info: SIL.Machine.AspNetCore.Services.SmtTransferEngineCommitService[0]
      SMT transfer engine commit service started.
2023-11-28 02:57:31	
machine-engine: 2023-11-28T07:57:31.950085029Z warn: Microsoft.AspNetCore.Server.Kestrel[0]
      Overriding address(es) 'http://+:80'. Binding to endpoints defined via IConfiguration and/or UseKestrel() instead.
2023-11-28 02:57:31	
serval-api: 2023-11-28T07:57:30.389080935Z fail: Serval.Shared.Controllers.ServiceUnavailableExceptionFilter[0]
      A user tried to access an unavailable service. See health-check logs for more details.
2023-11-28 02:57:31	
serval-api: 2023-11-28T07:57:30.387655249Z info: Grpc.Net.Client.Internal.GrpcCall[3]
      Call failed with gRPC error status. Status code: 'Unavailable', Message: 'Error connecting to subchannel.'.
2023-11-28 02:57:31	
serval-api: 2023-11-28T07:57:30.387533477Z fail: Grpc.Net.Client.Internal.GrpcCall[6]
      Error starting gRPC call.
      Grpc.Core.RpcException: Status(StatusCode="Unavailable", Detail="Error connecting to subchannel.", DebugException="System.Net.Sockets.SocketException: Connection refused")
       ---> System.Net.Sockets.SocketException (111): Connection refused
         at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken cancellationToken)
         at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.System.Threading.Tasks.Sources.IValueTaskSource.GetResult(Int16 token)
         at System.Net.Sockets.Socket.<ConnectAsync>g__WaitForConnectWithCancellation|277_0(AwaitableSocketAsyncEventArgs saea, ValueTask connectTask, CancellationToken cancellationToken)
         at Grpc.Net.Client.Balancer.Internal.SocketConnectivitySubchannelTransport.TryConnectAsync(ConnectContext context)
         --- End of inner exception stack trace ---
       
2023-11-28 02:57:31	
serval-api: 2023-11-28T07:57:28.354145404Z info: Serval.Shared.Controllers.ErrorResultFilter[0]
      Client ZVFeO8livsYcjG8P7zX7eDParKlaNRHN@clients made request:
       {
        "action": "TrainSegment",
        "controller": "TranslationEngines",
        "id": "6539938fe73fc85d889b0ebf",
        "version": "1"
      }.
       Serval responded with code 503. Trace: 00-3ce99279eaa728e79d93030d6f49feb2-356da3090cddc602-00
2023-11-28 02:57:29	
serval-api: 2023-11-28T07:57:28.350804443Z fail: Serval.Shared.Controllers.ServiceUnavailableExceptionFilter[0]
      A user tried to access an unavailable service. See health-check logs for more details.
2023-11-28 02:57:29	
serval-api: 2023-11-28T07:57:28.348960229Z info: Grpc.Net.Client.Internal.GrpcCall[3]
      Call failed with gRPC error status. Status code: 'Unavailable', Message: 'Error starting gRPC call. HttpRequestException: An error occurred while sending the request. IOException: The request was aborted. IOException: The response ended prematurely while waiting for the next frame from the server.'.
2023-11-28 02:57:29	
serval-api: 2023-11-28T07:57:28.347322419Z fail: Grpc.Net.Client.Internal.GrpcCall[6]
      Error starting gRPC call.
      System.Net.Http.HttpRequestException: An error occurred while sending the request.
       ---> System.IO.IOException: The request was aborted.
       ---> System.IO.IOException: The response ended prematurely while waiting for the next frame from the server.
         at System.Net.Http.Http2Connection.<ReadFrameAsync>g__ThrowMissingFrame|58_1()
         at System.Net.Http.Http2Connection.ReadFrameAsync(Boolean initialFrame)
         at System.Net.Http.Http2Connection.ProcessIncomingFramesAsync()
         --- End of inner exception stack trace ---
         at System.Net.Http.Http2Connection.ThrowRequestAborted(Exception innerException)
         at System.Net.Http.Http2Connection.Http2Stream.CheckResponseBodyState()
         at System.Net.Http.Http2Connection.Http2Stream.TryEnsureHeaders()
         at System.Net.Http.Http2Connection.Http2Stream.ReadResponseHeadersAsync(CancellationTo
2023-11-28 02:57:29	
serval-api: 2023-11-28T07:57:27.5036754Z info: System.Net.Http.HttpClient.SmtTransfer.ClientHandler[100]
      Sending HTTP request POST http://machine-engine/serval.translation.v1.TranslationEngineApi/TrainSegmentPair
2023-11-28 02:57:28	
machine-engine: 2023-11-28T07:57:10.362128764Z info: Grpc.AspNetCore.Server.ServerCallHandler[7]
      Error status code 'Aborted' with detail 'The engine must be built first.' raised.
2023-11-28 02:57:28	
serval-api: 2023-11-28T07:57:27.503632229Z info: System.Net.Http.HttpClient.SmtTransfer.LogicalHandler[100]
      Start processing HTTP request POST http://machine-engine/serval.translation.v1.TranslationEngineApi/TrainSegmentPair
2023-11-28 02:57:28	
serval-api: 2023-11-28T07:57:10.397540812Z info: Serval.Shared.Controllers.ErrorResultFilter[0]
      Client ZVFeO8livsYcjG8P7zX7eDParKlaNRHN@clients made request:
       {
        "action": "GetWordGraph",
        "controller": "TranslationEngines",
        "id": "6539938fe73fc85d889b0ebf",
        "version": "1"
      }.
       Serval responded with code 409. Trace: 00-a407832ca91de6e3863280a95fed2a8e-4d60dd22e7603fd8-00
2023-11-28 02:57:11	
serval-api: 2023-11-28T07:57:10.387553859Z info: Grpc.Net.Client.Internal.GrpcCall[3]
      Call failed with gRPC error status. Status code: 'Aborted', Message: 'The engine must be built first.'.
2023-11-28 02:57:11	
serval-api: 2023-11-28T07:57:10.385825257Z info: System.Net.Http.HttpClient.SmtTransfer.LogicalHandler[101]
      End processing HTTP request after 81.0251ms - 200
2023-11-28 02:57:11	
serval-api: 2023-11-28T07:57:10.385792177Z info: System.Net.Http.HttpClient.SmtTransfer.ClientHandler[101]
      Received HTTP response headers after 77.3348ms - 200
2023-11-28 02:57:11	
serval-api: 2023-11-28T07:57:10.29479925Z info: System.Net.Http.HttpClient.SmtTransfer.ClientHandler[100]
      Sending HTTP request POST http://machine-engine/serval.translation.v1.TranslationEngineApi/GetWordGraph
2023-11-28 02:57:11	
serval-api: 2023-11-28T07:57:10.293951424Z info: System.Net.Http.HttpClient.SmtTransfer.LogicalHandler[100]
      Start processing HTTP request POST http://machine-engine/serval.translation.v1.TranslationEngineApi/GetWordGraph
```

It appears that, while the engine is not built, it is in a corrupted state that is, although `engine.BuildRevision == 0`, when training a segment pair, it takes the poorly initilized engine and crashes the server.  This edge case should likely also be handled in Thot.

**Comment by ddaspit** [comment]
I don't see in the log where it is crashing the server.

**Comment by johnml1135** [comment]
There is no log where it crashes - I am assuming it was a segfault.  There is just a log for the missing service (and then a later log for the service restarting).

**Comment by ddaspit** [comment]
The engine is corrupted, but it doesn't cause a segmentation fault when it is loaded. I added code that still allows the engine to load. I tried training a segment pair, which is what seemed to cause the crash in this instance, and it did not cause a segmentation fault either. I cannot seem to reproduce the segmentation fault. I'm guessing that the low disk space caused the model in memory to get corrupted, which caused the segmentation fault. Did this crash occur once or over and over again as we saw in the other case of a corrupted model?

My plan right now is to make Thot more robust when loading a model with this kind of corruption. Hopefully, that will be sufficient to handle this kind of issue in the future.
